smarts/core/opendrive_road_network.py

```diff
@@ -0,0 +1,1810 @@
+    get_linear_segments_for_range,
+    offset_along_shape,
+    position_at_shape_offset,
+    vec_2d,
+    inplace_unwrap,
+    radians_to_vec,
+)
+
+from .lanepoints import LinkedLanePoint, LanePoints
+from .coordinates import BoundingBox, Point, Pose, RefLinePoint, Heading
+from smarts.core.utils.geometry import generate_mesh_from_polygons
+
+
+def _convert_camera(camera):
+    result = {
+        "name": camera.name,
+        "type": "perspective",
+        "perspective": {
+            "aspectRatio": camera.fov[0] / camera.fov[1],
+            "yfov": np.radians(camera.fov[1]),
+            "znear": float(camera.z_near),
+            # HACK: The trimesh gltf export doesn't include a zfar which Panda3D GLB
+            #       loader expects. Here we override to make loading possible.
+            "zfar": float(camera.z_near + 100),
+        },
+    }
+    return result
+
+
+gltf._convert_camera = _convert_camera
+
+
+class _GLBData:
+    def __init__(self, bytes_):
+        self._bytes = bytes_
+
+    def write_glb(self, output_path):
+        with open(output_path, "wb") as f:
+            f.write(self._bytes)
+
+
+@dataclass
+class LaneBoundary:
+    refline: PlanViewElement
+    inner: "LaneBoundary"
+    lane_widths: List[LaneWidthElement]
+    lane_offsets: List[LaneOffsetElement]
+    segment_size: float = 0.5
+
+    def refline_to_linear_segments(self, s_start: float, s_end: float) -> List[float]:
+        s_vals = []
+        geom_start = 0
+        for geom in self.refline._geometries:
+            geom_end = geom_start + geom.length
+            if type(geom) == LineGeometry:
+                s_vals.extend([geom_start, geom_end])
+            else:
+                s_vals.extend(
+                    get_linear_segments_for_range(
+                        geom_start, geom_end, self.segment_size
+                    )
+                )
+            geom_start = geom_start + geom.length
+        return [s for s in s_vals if s_start <= s <= s_end]
+
+    def get_lane_offset(self, s: float) -> float:
+        if len(self.lane_offsets) == 0:
+            return 0
+        i = 0
+        if s < self.lane_offsets[0].start_pos:
+            return 0
+        while i < len(self.lane_offsets) - 1:
+            if (
+                self.lane_offsets[i].start_pos
+                <= s
+                <= self.lane_offsets[i + 1].start_pos
+            ):
+                break
+            i += 1
+
+        poly = CubicPolynomial.from_list(self.lane_offsets[i].polynomial_coefficients)
+        ds = s - self.lane_offsets[i].start_pos
+        offset = poly.eval(ds)
+        return offset
+
+    def lane_width_at_offset(self, offset: float) -> LaneWidthElement:
+        i = 0
+        while i < len(self.lane_widths) - 1:
+            if (
+                self.lane_widths[i].start_offset
+                <= offset
+                <= self.lane_widths[i + 1].start_offset
+            ):
+                break
+            i += 1
```

Comment:
If `self.lane_widths` is already sorted by `start_offset` then we should use something like `bisect` here.